File: algo/47. Permutations II.js
// @9 DFS

/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var permuteUnique = function(nums) {
    var helper = function(res, nums, collection, size) {
        var collection = Array.from(collection);
        
        if (size == nums.length) {
            res.push(collection);
            return;
        }
        
        for (var i=0; i<nums.length; i++) {
                                   
            // on the same level, don't use the same number twice
            if (i>0 && nums[i] == nums[i-1] && !visited[i-1]) {
                continue;
            }
            
            // if we have visisted this node
            if (visited[i]) {
                continue;
            }
                        
            collection.push(nums[i]);
            visited[i] = 1;
            helper(res, nums, collection, size+1);
            collection.pop();
            visited[i] = 0;
        }
    };
       
    var res = [];
    
    if (nums.length == 0) {
        return res;
    }
    
    // use an array to store if a specific node has been visited
    var visited = [];
    for (var i=0; i<nums.length; i++) {
        visited.push(0);
    }
    
    nums.sort();
    helper (res, nums, [], 0);
    return res;
};

File: round2/47. Permutations II.cs
public class Solution {
    private void DFS(List<IList<int>> res, List<int> candidate, int[] visited, int[] nums, int level, int size) {
        if (level == size) {
            res.Add(new List<int>(candidate));
            return;
        }
        for (int i=0; i<size; i++) {
            if (visited[i] == 1) continue;
            if (i>0 && nums[i]==nums[i-1] && visited[i-1]==1) continue;
            visited[i] = 1;
            candidate.Add(nums[i]);
            DFS(res, candidate, visited, nums, level+1, size);
            candidate.RemoveAt(candidate.Count-1);
            visited[i] = 0;
        }
    }
    public IList<IList<int>> PermuteUnique(int[] nums) {
        int size = nums.Length;
        var res = new List<IList<int>>();
        if (size==0) return res;
        Array.Sort(nums);
        var candidate = new List<int>();
        int[] visited = new int[size];
        DFS(res, candidate, visited, nums, 0, size);
        return res;
    }
}